relm4-macros/src/attrs.rs

```diff
@@ -1,31 +1,176 @@
+    /// ```
+    /// is illegal
+    relm4_path_set: bool,
 }
 
-impl Parse for Attrs {
-    fn parse(input: ParseStream) -> Result<Self> {
-        let pub_vis = if input.is_empty() {
-            None
-        } else {
-            Some(input.parse()?)
+impl Attrs {
+    fn new() -> Self {
+        let relm4_path_segment = PathSegment {
+            ident: Ident::new("relm4", Span::call_site()),
+            arguments: PathArguments::None,
         };
+        
+        let mut relm4_segments: Punctuated<PathSegment, Colon2> = Punctuated::new();
+        relm4_segments.push(relm4_path_segment);
 
-        Ok(Attrs { pub_vis })
+        let relm4_path: Path = Path {
+            leading_colon: Some(Token![::](Span::call_site())),
+            segments: relm4_segments,
+        };
+
+        Attrs {
+            visibility: None,
+
+            relm4_path,
+            relm4_path_set: false,
+        }
     }
 }
 
-impl ToTokens for Attrs {
-    fn to_tokens(&self, out: &mut TokenStream2) {
-        if let Some(vis_pub) = &self.pub_vis {
-            vis_pub.to_tokens(out);
+impl Parse for Attrs {
+    /// Rules for parsing attributes
+    /// 
+    /// 1. It's fine if visibility is used unnamed so `#[widget(pub)]` must be valid but thats the only case
+    /// 2. Widget visibility might be named `#[widget(visibility = pub)]`
+    /// 3. `relm4` argument must be named. Always
+    ///
+    fn parse(input: ParseStream) -> Result<Self> {
+        let mut attrs = Attrs::new();
+        let mut attrs_type = AttributeType::None;
+        let mut dangling_comma;
+
+        let mixed_use_error_message = 
+            "You can't mix named and unnamed arguments while using `relm4_macros::widget`. \n\
+            \n\
+            You can use one of\n\
+            \n\
+            1. `#[relm4_macros::widget()]` to define widget with private visibility\n\
+            2. `#[relm4_macros::widget(pub)]` to define widget with public visibility\n\
+            3. `#[relm4_macros::widget( visibility = pub )]` to define widget with public visibility and potentially other arguments\n\
+            \n\
+            Please use `visibility = pub` to fix this error";
+
+        while !input.is_empty() {            
+            dangling_comma = true;
+
+            if input.peek(Token![pub]) {
+                if matches!(attrs_type, AttributeType::Named) {
+                    return Err(input.error(mixed_use_error_message));
+                }
+                if attrs.visibility.is_some() {
+                    return Err(input.error("You can't assign visibility twice"));
+                }
+                let pub_vis: Visibility = input.parse()?;
+
+                attrs_type = AttributeType::Unnamed{
+                    span: pub_vis.span(),
+                };
+                attrs.visibility = Some(pub_vis);
+                dangling_comma = false;
+            }
+            else if input.peek(Ident) && input.peek2(Token![=]){
```

Comment:
I would rather just parse `Ident` and the `=` unconditionally. Then you don't need `dangling_comma` and the user gets an error like `expected identifier` automatically.